src/vncproxy.ml

```diff
@@ -121,43 +119,46 @@ let bind_local_port () =
 
 let connect_to_console session_id console =
   let uri = Uri.of_string console.API.console_location in
-  match%lwt Xen_api_lwt_unix.Lwt_unix_IO.open_connection uri with
+  Xen_api_lwt_unix.Lwt_unix_IO.open_connection uri >>= function
   | Ok ((_, ic), (_, oc)) ->
-    let%lwt () = Lwt_io.write oc (Printf.sprintf "CONNECT %s?%s HTTP/1.0\r\n" (Uri.path uri) (Uri.(encoded_of_query (query uri)))) in
-    let%lwt () = Lwt_io.write oc (Printf.sprintf "Cookie: session_id=%s\r\n" (API.Ref.string_of session_id)) in
-    let%lwt () = Lwt_io.write oc "\r\n" in
-    let%lwt () = Lwt_io.flush oc in
-    let%lwt status = Lwt_io.read_line ic in
+    Lwt_io.write oc (Printf.sprintf "CONNECT %s?%s HTTP/1.0\r\n" (Uri.path uri) (Uri.(encoded_of_query (query uri)))) >>= fun () ->
+    Lwt_io.write oc (Printf.sprintf "Cookie: session_id=%s\r\n" (API.Ref.string_of session_id)) >>= fun () ->
+    Lwt_io.write oc "\r\n" >>= fun () ->
+    Lwt_io.flush oc >>= fun ()->
+    Lwt_io.read_line ic >>= fun status ->
     Printf.fprintf stderr "%s\n%!" status;
     let rest = ref [] in
     let finished = ref false in
-    let%lwt () = while%lwt not !finished do
-        let%lwt line = Lwt_io.read_line ic in
-        Printf.fprintf stderr "%s\n%!" line;
-        if line = ""
-        then finished := true
-        else rest := line :: !rest;
-        return ()
-      done in
-    return (ic, oc)
+    let rec loop () =
+      if not !finished then
+        (Lwt_io.read_line ic >>= fun line ->
+         Printf.fprintf stderr "%s\n%!" line;
+         if line = "" then finished := true
+         else rest := line :: !rest;
+         return ()
+
+        ) >>= loop
+      else Lwt.return_unit
```

Comment:
`rest` is not used anywhere, I think is a leftover of some old implementation. 
We can also get rid of the mutation in finished (I think) with something like the following
```ocaml
let rec loop finished () = 
  match finished with
  | true -> Lwt.return_unit
  | false -> begin
    Lwt_io.read_line ic >>= fun line ->
    Printf.fprintf stderr "%s\n%!" line;
    Lwt.return_unit
    >>= loop (line = "")
  end
in
loop false () >>= fun () -> 
return (ic, oc)
```

or using `if ... then ... else`.
It may even be further simplified removing the `return_unit` step and directly calling `loop` there:
```ocaml
let rec loop finished = 
  match finished with
  | true -> Lwt.return_unit
  | false ->
    Lwt_io.read_line ic >>= fun line ->
    Printf.fprintf stderr "%s\n%!" line;
    loop (line = "")
in
loop false >>= fun () -> 
return (ic, oc)
```
But it probably would be a good idea to test it and double check.

Reply:
I'm not sure 'finished' can be removed like this, because 'line' is local to that function, it has to be returned somehow so it can be bound to loop.